### PR TITLE
feat: add additional process exec functions

### PR DESCRIPTION
**Pull request by varl**
This is a port of the utils/run.js file from cli-style:
https://github.com/dhis2/cli-style/blob/2bfb5ef5ee2b26ff7e27594fe36cef9eb12b2360/src/utils/run.js

It introduces cross-spawn as the default way to spawn processes which
works well on other platforms (e.g. windows) and exposes additional
functions to spawn and use the result of processes.

Notably, it adds the bin() function that resolves a command to the
nearest node_modules/.bin folder based on the cwd.

This is needed for cli-style, but also now needed for cli-utils-cypress,
so makes sense to generalize it here.

### CODE REVIEW

File: lib/exec.js [hunk 1]
```diff
@@ -66,3 +68,82 @@ module.exports = function ({
...
+    }
+}
+
+function exit (code, msg) {
+    if (msg && code > 0) {
+        reporter.print('')
+        reporter.error(msg)
+    }
+    process.exit(code)
+}
```
Comment (amcgee) [inline lib/exec.js:125]: I like having this function!  Should it maybe live somewhere else though?  Seems a bit unrelated to the execution of child processes?

File: lib/exec.js [hunk 2]
```diff
@@ -66,3 +68,82 @@ module.exports = function ({
...
+    if (result.error) {
+        throw result.error
+    }
+
+    if (callback) {
+        return callback(result)
+    }
+
+    if (result.status !== 0) {
+        process.exit(result.status === null ? 0 : result.status)
```
Comment (amcgee) [inline lib/exec.js:137]:
```suggestion
        exit(result.status === null ? 0 : result.status)
```

**Review by amcgee — APPROVED** [review]
🎉  huzzah, this is great!  Backwards-compatible and adds a bunch of useful stuff.  Added a couple minor comments.

**Review by amcgee — CHANGES_REQUESTED** [review]
Actually, I think these should also be re-exported from `index.js` (root level)

File: lib/exec.js (hunk 1 above)
Comment (varl) [reply lib/exec.js:125]: Yeah, it is a bit unrelated. We could have a "process" module that wraps process stuff. Another case would be to wrap the `process.cwd()` command to allow setting a custom `cwd` instead of rolling a custom solution for that in each tool.

**Comment by varl** [comment]
> Actually, I think these should also be re-exported from `index.js` (root level)

They are, but no changes necessary.

```node
> const exec = require('./index.js').exec

> console.dir(exec)
[Function: exec] {
  exit: [Function: exit],
  spawn: [Function: spawn],
  run: [Function: run],
  bin: [Function: bin],
  callback: [Function: callback]
}
```

So `exec` (the default) is the `exec` function as it was before.

But we can also do named imports:

```node
> const { spawn, run } = require('./index.js').exec

> console.dir(spawn)
[Function: spawn]

> console.dir(run)
[Function: run]
```

Made it more explicit in b422dde instead of relying on Node 14+ functionality.

File: lib/exec.js (hunk 1 above)
Comment (varl) [reply lib/exec.js]: 8d973fc

**Review by amcgee — COMMENTED** [review]
👌 - should we name `process` something else so it doesn't inadvertently override the global?  Maybe `processHelpers`...?  Not sure...

```
const { process } = require('@dhis2/cli-helpers-engine')
console.log(process.env)
// undefined
```

**Comment by varl** [comment]
> ok_hand - should we name `process` something else so it doesn't inadvertently override the global? Maybe `processHelpers`...? Not sure...
> 
> ```js
> const { process } = require('@dhis2/cli-helpers-engine')
> console.log(process.env)
> // undefined
> ```

Hm, sure. I mostly use the form, so I didn't consider it a problem:

```
const { log } = require('@dhis2/cli-helpers-engine').reporter
const { exit } = require('@dhis2/cli-helpers-engine').process
```

I don't like "*Helpers" because it's already in a package named "helpers" so it's a bit redundant.

**Comment by varl** [comment]
`exit` is a fundamental operation, so best place I can think of that doesn't feel awkward is straight from the helpers-engine package:

```node
const { exit } = require('@dhis2/cli-helpers-engine')

const engine = require('@dhis2/cli-helpers-engine')
engine.exit()
```

**Review by amcgee — APPROVED** [review]
(we have [quite](https://github.com/dhis2/app-platform/blob/master/cli/src/commands/deploy.js#L2) [a](https://github.com/dhis2/cli/blob/master/packages/utils/src/cmds/schema/fetch.js#L7) [few](https://github.com/dhis2/app-platform/blob/master/cli/src/lib/compiler/compile.js#L3) [cases](https://github.com/dhis2/cli/blob/master/packages/cluster/src/commands/list.js#L1) of `const { reporter } = require('@dhis2/cli-helpers-engine')` so I think it's nice to support that)

We could maybe consider re-exporting all the exec methods, but I think this is fine for now 👌 

🎉  looks good to me!
